### PR TITLE
[#2498] feat(spark-connector): support complex types for spark-connector

### PR DESCRIPTION
**Pull request by caican00**
### What changes were proposed in this pull request?

support complex types for spark-connector.

### Why are the changes needed?

Complex types havn't been supported in spark-connector before this PR.

Fix: https://github.com/datastrato/gravitino/issues/2498

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
New UT and IT.

### CODE REVIEW

**Comment by caican00** [comment]
Hi @FANNG1 Could you help review this PR? Thank you.

**Comment by FANNG1** [comment]
cool

**Comment by FANNG1** [comment]
could you add complex type test in `SparkIT`?

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/SparkTypeConverter.java [hunk 1]
```diff
@@ -101,6 +131,30 @@ public static DataType toSparkType(Type gravitinoType) {
...
+      List<StructField> fields =
+          Arrays.stream(structType.fields())
+              .map(
+                  f ->
+                      DataTypes.createStructField(
+                          f.name(),
+                          toSparkType(f.type()),
+                          f.nullable(),
+                          new MetadataBuilder()
+                              .putString(ConnectorConstants.COMMENT, f.comment())
```
Comment (FANNG1) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/SparkTypeConverter.java:154]: should we process f.comment == null?

File: spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java [hunk 2]
```diff
@@ -38,6 +41,16 @@
 public class TestSparkTypeConverter {
 
   private HashMap<Type, DataType> gravitinoToSparkTypeMapper = new HashMap<>();
+  private StructType gravitinoStructType =
```
Comment (FANNG1) [inline spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java:44]: could you make struct type test cover more cases? like nested struct type, cover different column comment, nullable

**Comment by caican00** [comment]
> could you add complex type test in `SparkIT`?

sure, i will do it.

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/SparkTypeConverter.java (hunk 1 above)
Comment (caican00) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/SparkTypeConverter.java:154]: i'll think it over.

File: spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java (hunk 2 above)
Comment (caican00) [reply spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java:44]: sure

**Comment by caican00** [comment]
> > could you add complex type test in `SparkIT`?
> 
> sure, i will do it.

added test case in `SparkIT`.

File: spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java (hunk 2 above)
Comment (caican00) [reply spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java:46]: updated.

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/SparkTypeConverter.java (hunk 1 above)
Comment (caican00) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/SparkTypeConverter.java]: Fixed.

**Comment by caican00** [comment]
all comments have been addressed. cc @FANNG1

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/SparkTypeConverter.java [hunk 3]
```diff
@@ -101,6 +131,32 @@ public static DataType toSparkType(Type gravitinoType) {
     } else if (gravitinoType instanceof Types.TimestampType
         && !((Types.TimestampType) gravitinoType).hasTimeZone()) {
       return DataTypes.TimestampNTZType;
+    } else if (gravitinoType instanceof Types.ListType) {
+      Types.ListType listType = (Types.ListType) gravitinoType;
+      return ArrayType.apply(toSparkType(listType.elementType()), listType.elementNullable());
```
Comment (FANNG1) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/SparkTypeConverter.java:136]: `DataTypes` provides helper method to create Array and Map, could you use helper methods?

File: spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java [hunk 4]
```diff
@@ -59,6 +110,98 @@ void init() {
     gravitinoToSparkTypeMapper.put(DateType.get(), DataTypes.DateType);
     gravitinoToSparkTypeMapper.put(TimestampType.withTimeZone(), DataTypes.TimestampType);
     gravitinoToSparkTypeMapper.put(TimestampType.withoutTimeZone(), DataTypes.TimestampNTZType);
+    gravitinoToSparkTypeMapper.put(
+        ListType.of(IntegerType.get(), true), DataTypes.createArrayType(DataTypes.IntegerType));
+    gravitinoToSparkTypeMapper.put(
+        MapType.of(IntegerType.get(), StringType.get(), true),
+        DataTypes.createMapType(DataTypes.IntegerType, DataTypes.StringType));
```
Comment (FANNG1) [inline spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java:117]: could you put struct in `gravitinoToSparkTypeMapper` too? then no need to do extral check for struct

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java [hunk 5]
```diff
@@ -405,6 +405,31 @@ void testAlterTableUpdateColumnComment() {
...
   }
 
+  @Test
+  void testComplexType() {
+    String tableName = "test_complex_type";
+    dropTableIfExists(tableName);
+
+    sql(
+        String.format(
+            "CREATE TABLE %s (col1 ARRAY<int> COMMENT 'array', col2 MAP<string, string> COMMENT 'map', col3 STRUCT<col1: int, col2: string> COMMENT 'struct')",
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:415]: change `int` `string` to upper case?

Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:415]: make Map key and value different types?

Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java]: fixed.

Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java]: fixed.

File: spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java (hunk 4 above)
Comment (caican00) [reply spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java:70]: done

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/SparkTypeConverter.java (hunk 3 above)
Comment (caican00) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/SparkTypeConverter.java]: fixed.

**Comment by caican00** [comment]
all new comments have been addressed.

File: spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java [hunk 6]
```diff
@@ -59,6 +63,54 @@ void init() {
...
+    gravitinoToSparkTypeMapper.put(
+        StructType.of(
```
Comment (FANNG1) [inline spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java:72]: could you provide a function to create spark and gravitino types to make it more readable?

Comment (FANNG1) [reply spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java:72]: like Pair<GravitinoStruct, SparkStruct> createTestStructTypes()

File: spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java [hunk 7]
```diff
@@ -59,6 +63,60 @@ void init() {
     gravitinoToSparkTypeMapper.put(DateType.get(), DataTypes.DateType);
     gravitinoToSparkTypeMapper.put(TimestampType.withTimeZone(), DataTypes.TimestampType);
     gravitinoToSparkTypeMapper.put(TimestampType.withoutTimeZone(), DataTypes.TimestampNTZType);
+    gravitinoToSparkTypeMapper.put(
+        ListType.of(IntegerType.get(), true), DataTypes.createArrayType(DataTypes.IntegerType));
+    gravitinoToSparkTypeMapper.put(
+        MapType.of(IntegerType.get(), StringType.get(), true),
+        DataTypes.createMapType(DataTypes.IntegerType, DataTypes.StringType));
+    gravitinoToSparkTypeMapper.put(createGravitinoStructType(), createSparkStructType());
```
Comment (caican00) [inline spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java:71]: How about this? create static method for gravitino and spark to build their structType. cc @FANNG1

Comment (FANNG1) [reply spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java:71]: ok, please provide some description to the two functions, and move private functions after public functions.

**Comment by FANNG1** [comment]
LGTM except minor comments, @jerryshao @qqqttt123 @mchades  do you have time to review?

File: spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java (hunk 7 above)
Comment (caican00) [reply spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java:71]: done.

File: spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java (hunk 6 above)
Comment (caican00) [reply spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/TestSparkTypeConverter.java]: Fixed.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java [hunk 8]
```diff
@@ -405,6 +405,33 @@ void testAlterTableUpdateColumnComment() {
     checkTableColumns(tableName, updateCommentColumns, getTableInfo(tableName));
   }
 
+  @Test
+  void testComplexType() {
+    String tableName = "test_complex_type";
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:410]: please rename to `complex_type_table`;

Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:410]: ok

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java [hunk 9]
```diff
@@ -405,6 +405,33 @@ void testAlterTableUpdateColumnComment() {
...
+                DataTypes.createMapType(DataTypes.StringType, DataTypes.IntegerType),
+                "map"),
+            SparkColumnInfo.of(
+                "col3",
+                DataTypes.createStructType(
+                    Arrays.asList(
+                        DataTypes.createStructField("col1", DataTypes.IntegerType, true),
+                        DataTypes.createStructField("col2", DataTypes.StringType, true))),
+                "struct"));
+    checkTableColumns(tableName, expectedSparkInfo, tableInfo);
```
[the remark this answers is not in the document]
Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:432]: sure

[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:432]: you should add complex type to `typeConstant` to generate test data

[the remark this answers is not in the document]
Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:445]: have fixed the IT. Could you help review this again? cc @FANNG1

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java [hunk 10]
```diff
@@ -431,8 +473,22 @@ private void checkTableReadWrite(SparkTableInfo table) {
             .map(Object::toString)
             .map(
                 s -> {
-                  String tmp = org.apache.commons.lang3.StringUtils.removeEnd(s, "'");
-                  tmp = org.apache.commons.lang3.StringUtils.removeStart(tmp, "'");
+                  String tmp = org.apache.commons.lang3.StringUtils.remove(s, "'");
+                  if (org.apache.commons.lang3.StringUtils.isEmpty(tmp)) {
+                    return tmp;
+                  } else if (tmp.startsWith("array")) {
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:479]: could you add some description about why you do this?  better to provide some example

Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:483]: updated.

**Comment by caican00** [comment]
Hi @FANNG1 could you help re-trigger this pipeline for what appears to be an internal error?

**Comment by FANNG1** [comment]
> Hi @FANNG1 could you help re-trigger this pipeline for what appears to be an internal error? And i can't seem to re-trigger any pipeline. Thank you very much.

done

**Comment by caican00** [comment]
> > Hi @FANNG1 could you help re-trigger this pipeline for what appears to be an internal error? And i can't seem to re-trigger any pipeline. Thank you very much.
> 
> done

run successfully.

**Review by FANNG1 — APPROVED** [review]

**Comment by FANNG1** [comment]
@caican00 , thanks for your work !

**Comment by caican00** [comment]
> @caican00 , thanks for your work !

@FANNG1 , thanks for review and merge.

**Comment by FANNG1** [comment]
@caican00 , some types are not supported yet, like `NullType`, `IntervalType` do you like continue work on it?

**Pull request merged**